certora/KilnBase.spec

```diff
@@ -0,0 +1,195 @@
+    lot() returns (uint256) envfree
+    hop() returns (uint256) envfree
+    zzz() returns (uint256) envfree
+    locked() returns (uint256) envfree
+    sell() returns (address) envfree
+    buy() returns (address) envfree
+    dai.totalSupply() returns (uint256) envfree
+    dai.balanceOf(address) returns (uint256) envfree
+}
+
+definition WAD() returns uint256 = 10^18;
+
+ghost lockedGhost() returns uint256;
+
+hook Sstore locked uint256 n_locked STORAGE {
+    havoc lockedGhost assuming lockedGhost@new() == n_locked;
+}
+
+hook Sload uint256 value locked STORAGE {
+    require lockedGhost() == value;
+}
+
+// Verify fallback always reverts
+rule fallback_revert(method f) filtered { f -> f.isFallback } {
+    env e;
+
+    calldataarg arg;
+    f@withrevert(e, arg);
+
+    assert(lastReverted, "Fallback did not revert");
+}
+
+// Verify that wards behaves correctly on rely
+rule rely(address usr) {
+    env e;
+
+    address other;
+
+    require(other != usr);
+
+    uint256 wardOtherBefore = wards(other);
+
+    rely(e, usr);
+
+    uint256 wardAfter = wards(usr);
+    uint256 wardOtherAfter = wards(other);
+
+    assert(wardAfter == 1, "rely did not set wards as expected");
+    assert(wardOtherAfter == wardOtherBefore, "rely affected other wards which was not expected");
+}
+// Verify revert rules on rely
+rule rely_revert(address usr) {
+    env e;
+
+    uint256 ward = wards(e.msg.sender);
+
+    rely@withrevert(e, usr);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = ward != 1;
+
+    assert(revert1 => lastReverted, "revert1 failed");
+    assert(revert2 => lastReverted, "revert2 failed");
+
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify that wards behaves correctly on deny
+rule deny(address usr) {
+    env e;
+
+    address other;
+
+    require(other != usr);
+
+    uint256 wardOtherBefore = wards(other);
+
+    deny(e, usr);
+
+    uint256 wardAfter = wards(usr);
+    uint256 wardOtherAfter = wards(other);
+
+    assert(wardAfter == 0, "deny did not set wards as expected");
+    assert(wardOtherAfter == wardOtherBefore, "deny affected other wards which was not expected");
+}
+// Verify revert rules on deny
+rule deny_revert(address usr) {
+    env e;
+
+    uint256 ward = wards(e.msg.sender);
+
+    deny@withrevert(e, usr);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = ward != 1;
+
+    assert(revert1 => lastReverted, "revert1 failed");
+    assert(revert2 => lastReverted, "revert2 failed");
+
+    assert(lastReverted => revert1 || revert2, "Revert rules are not covering all the cases");
+}
+
+// Verify correct storage changes for not reverting file
+rule file(bytes32 what, uint256 data) {
+    env e;
+
+    uint256 lotBefore = lot();
+    uint256 hopBefore = hop();
+
+    file(e, what, data);
+
+    assert(what == 0x6c6f740000000000000000000000000000000000000000000000000000000000 => lot() == data, "file did not set lot as expected");
```

Comment:
It is always better chaching the values in variabels (aka use `lotAfter`). In case of an issue if the variable exists will show up in the debug, otherwise not.

Reply:
good point, done in 58556c6